### ISSUE TITLE
Support for Resource Instances within Azure Storage Networking

### ISSUE THREAD

**Issue by Lachlan-White** [terraform-providers/terraform-provider-azurerm#10965]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
In the case of utilising an Azure Synapse Workspace within a Managed Virtual Network, to connect this Workspace to a secured Azure Data Lake or Storage account the use of a Resource Instance within the Azure Storage Account Networking Blade can be utilised.

This capability is in public preview but is soon to be a recommended way in which to implement connectivity between Microsoft Managed Virtual Networks and the Managed identity from within your tenant.
<https://docs.microsoft.com/en-us/azure/storage/common/storage-network-security?tabs=azure-portal#trusted-access-based-on-system-assigned-managed-identity>

### New or Affected Resource(s)

* `azurerm_storage_account`
* `azurerm_stoage_account_netwrok_rules`

### Potential Terraform Configuration

```hcl
resource "azurerm_storage_account" "example" {
  name                = "storageaccountname"
  resource_group_name = azurerm_resource_group.example.name

  location                 = azurerm_resource_group.example.location
  account_tier             = "Standard"
  account_replication_type = "LRS"

  network_rules {
    default_action             = "Deny"
    ip_rules                   = ["100.0.0.1"]
    virtual_network_subnet_ids = [azurerm_subnet.example.id]
    resource_instance_type = "Microsoft.Synapse/workspaces"
   resource_instance_name = "mysynapseworkspace"
  }
```

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor blog posts or documentation? For example:

*https://docs.microsoft.com/en-us/azure/synapse-analytics/security/connect-to-a-secure-storage-account#grant-your-azure-synapse-workspace-access-to-your-secure-storage-account-as-a-trusted-azure-service
* https://docs.microsoft.com/en-us/azure/storage/common/storage-network-security?tabs=azure-portal#grant-access-from-azure-resource-instances-preview

**Comment by drjwelch** [comment]
Same applies for other services, in our case data factory.  For anyone else arriving here, one workaround is to use self-hosted IR and allow its subnet through the firewall.

**Comment by a30000931** [comment]
over a year since being raised  - any news on this?

**Comment by WaitingForGuacamole** [comment]
Agreed, an update would be useful - closed issue https://azure.microsoft.com/en-us/updates/public-preview-ephemeral-os-disks-supports-hostbased-encryption-using-customer-managed-key/ is referenced above, which just links back to this issue, so it's hard to know if this issue is in the backlog?

**Comment by loczek09** [comment]
Hi,
There is private_link_access block within  network_rules block where you can define id of the to grant access.
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account#network_rules

**Comment by mahalel** [comment]
I can confirm @loczek09's solution works:

```hcl
resource "azurerm_storage_account_network_rules" "this" {

  storage_account_id = azurerm_storage_account.this.id

  default_action             = "Deny"
  ip_rules                   = []
  virtual_network_subnet_ids = []
  bypass                     = ["AzureServices"]
  private_link_access {
    endpoint_resource_id       = module.synapse_workspace[0].id
  }
}
```

**Comment by kimdogbe** [comment]
This a good solution for a single resource with an id. I think this is still an issue when trying to set the "Instance name" value to "All in current subscription" or "All in current tenant" like so: 
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/25349773/a21917f8-acc6-46ce-acec-35d2cb3ad908)

**Comment by mikewoodd3432** [comment]
> This a good solution for a single resource with an id. I think this is still an issue when trying to set the "Instance name" value to "All in current subscription" or "All in current tenant" like so: ![image](https://user-images.githubusercontent.com/25349773/259369223-a21917f8-acc6-46ce-acec-35d2cb3ad908.png)
> 
> Anyone know if there is a way to setup in terraform to get the results above?

@kimdogbe I had a play around with configuring this on a resource and then looking at the Resource Explorer. For all Synapse Workspaces in current resource group the endpoint resource id might be something like "<resource_group_id>/providers/Microsoft.Synapse/workspaces/*"

**Comment by ikhudur** [comment]
Hello,

It has been almost 3 years since this has been raised. Any updated on if/when this will be introduced?

I want to allow a Fabric workspace to access a storage account as explained in [1] and [2], but currently (as I know) it is not possible to create a resource instance rule for resources outside of Azure?

[1] https://blog.fabric.microsoft.com/sv-se/blog/introducing-trusted-workspace-access-for-onelake-shortcuts?ft=All
[2] https://learn.microsoft.com/en-us/fabric/security/security-trusted-workspace-access

**Comment by JoshWeepie** [comment]
Yup, just ran into this as well, I wanted to add Data Factory, but with the 'All in current subscription'.

**Comment by Tanchwa** [comment]
Does anyone know if the workaround currently supports Azure VM managed identities as an argument? We have multiple SHIRs for different ADF workloads, and want to make sure only one of those instances can reach a particular storage account. Both SHIRs currently reside in the same subnet.

**Comment by CaishenNefri** [comment]
I think this is already implemented via private_link_access
https://github.com/hashicorp/terraform-provider-azurerm/issues/14816#issuecomment-1006364592

**Comment by holgerjs** [comment]
> I think this is already implemented via private_link_access [#14816 (comment)](https://github.com/hashicorp/terraform-provider-azurerm/issues/14816#issuecomment-1006364592)

That's correct. It's possible to use wildcards for the `endpoint_resource_id` property and thereby allow all resources of a certain type, i.E.:

```    
private_link_access {
      endpoint_tenant_id   = data.azurerm_client_config.current.tenant_id
      endpoint_resource_id = "/subscriptions/${data.azurerm_client_config.current.subscription_id}/resourceGroups/*/providers/Microsoft.MachineLearningServices/workspaces/*"
}
```
